app/components/sorted_table_header_link_component.rb

```diff
@@ -0,0 +1,104 @@
+#
+# This class does *not* take care of searching or filtering the table;
+# it's only in charge of creating the links.
+#
+# Controller method:
+#   @link_maker = SortedTableHeaderLinkComponent.link_maker(
+#     params:               collection_params, # the controller's permitted ActiveRecord params
+#     table_sort_field_key: :sort_field,       # keys for looking up the table's current sort field in the above
+#     table_sort_order_key: :sort_order,       # keys for looking up the table's current sort order (asc or desc)
+#   )
+#
+#
+# Now, for each sortable column in the template, call @link_maker.link with the column info to render the actual title link:
+#   render(@link_maker.link(column_title: "title",         sort_field: "title"     ))
+#   render(@link_maker.link(column_title: "Created",       sort_field: "created_at"))
+#   render(@link_maker.link(column_title: "Last Modified", sort_field: "updated_at"))
+class SortedTableHeaderLinkComponent < ApplicationComponent
+
+  # @param [ActionController::Parameters] params the permitted params from the controller
+  # @param [Symbol] table_sort_field_key the current sort field of the table.
+  # @param [Symbol] table_sort_order_key the current sort order of the table.
+  # @return [SortedTableHeaderLinkComponent::LinkMaker] a LinkMaker object whose link method returns a SortedTableHeaderLinkComponent with prepopulated fields.
+  def self.link_maker(params:, table_sort_field_key:, table_sort_order_key:)
+    raise ArgumentError.new("Unpermitted params") unless params.permitted?
+    self::LinkMaker.new(
+      params: params,
+      table_sort_field_key: table_sort_field_key,
+      table_sort_order_key: table_sort_order_key
+    )
+  end
+
+  # @param [String] column_title The text of the link; the column title as displayed to the user.
+  # @param [String] sort_field the new sort field that will be used if a user clicks the link.
+  # @param [String] table_sort_field the current sort field of the table.
+  # @param [String] table_sort_order the current sort order of the table.
+  # @param [Hash] other_params any and all extra paramaters to add to the link.
+  def initialize(column_title:, sort_field:, table_sort_field:, table_sort_order:, other_params:)
+    unless column_title.present? && sort_field.present?
+      raise ArgumentError.new("Missing arguments: column_title or sort field") 
+    end
+    unless table_sort_order.in? ['asc', 'desc']
+      raise ArgumentError.new("Sort order needs to be asc or desc") 
+    end
+    @column_title = column_title
+    @sort_field = sort_field
+    @table_sort_field = table_sort_field
+    @table_sort_order = table_sort_order
+    @other_params = other_params
+  end
+
+  def call
+    link_to sort_column_title, sort_column_path
+  end
+
+  def sort_column_path
+    url_for ( {
+        sort_field:   @sort_field,
+        sort_order:   (reversed_sort_order if @sort_field == @table_sort_field)
+      }.merge!(@other_params)
```

Comment:
`sort_field` and `sort_order` as param keys are hard-coded here. Can we make them arguments to the initializer instead, so one page could call it a different thing like `search_order` and `search_direction` or whatever ?  You can still have them have defaults, optional args.

Reply:
Also, use rails `reverse_merge!` or change the order of what's merged into what, so if @other_params has conflicting `sort_field` and `sort_order` keys, our new intentional ones are not overwritten.

Reply:
The line `(reversed_sort_order if @sort_field == @table_sort_field)` is confusing to me, although I understand what the intention is. 

Your'e leaving out `sort_order` value UNLESS it's reversed, otherwise it's blank? Is this advisable/what ransack did?  If it does have a blank value and that's intentional, let's make sure it doesn't have an unneeded key in the url either with a blank value?

I think it might be more straightforward to put something in the constructor or even the link_maker that decides what the direction shoudl be and passes is as an arg or stores it in a variable called `@direction` or `@order_value` or something, that's present whether or not it's reversed?  I'm just finding this logic confusing, although it very well may be correct!